### PR TITLE
[BEAM-6350] Reuse PCollectionView when created in translators

### PR DESCRIPTION
**Pull request by mareksimunek**
If for LeftJoin is used BroadcastHashJoinTranslator then from right side is created PCollectionView (as sideInput).

If we use right side in multiple joins then PCollectionView  is created multiple times which is not optimal behavior.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by dmvk** [comment]
Run JavaPortabilityApi PreCommit

File: sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/PViewsStore.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+   *
+   * @param key mapping key
+   * @param pCollectionToView a {@link PCollection} view will be created from by applying {@link
+   *     View#asMultimap()}
+   * @param <K> element key type
+   * @param <V> value key type
+   * @return the current (already existing or computed) value associated with the specified key
+   */
+  @SuppressWarnings("unchecked")
+  public <K, V> PCollectionView<Map<K, Iterable<V>>> computeViewAsMultimapIfAbsent(
```
Comment (dmvk) [inline sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/PViewsStore.java]: can we have a more readable name? eg.: `asMultimap`

File: sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/EuphoriaOptions.java [hunk 2]
```diff
@@ -60,4 +70,13 @@ public TranslatorProvider create(PipelineOptions options) {
...
   void setAccumulatorProviderFactory(AccumulatorProvider.Factory accumulatorProviderFactory);
+
+  @Description("Map of known PCollectionViews")
+  @Default.InstanceFactory(DefaultPCollectionViewStoreFactory.class)
+  @JsonIgnore
+  @Internal
+  PViewsStore getPCollectionViewsStore();
+
+  @Internal
+  void setPCollectionViewsStore(PViewsStore pCollectionViewsStore);
```
Comment (dmvk) [inline sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/EuphoriaOptions.java]: is setter necessary?

Comment (dmvk) [reply sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/EuphoriaOptions.java]: I reckon we don't need a separate field in pipeline options. It should be enough to keep instance property in BHJ Translator.

File: sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/BroadcastHashJoinTranslator.java [hunk 3]
```diff
@@ -42,15 +41,24 @@
...
+    // And therefore multiple broadcasts of the same data.
+    final PViewsStore pViews =
+        left.getPipeline().getOptions().as(EuphoriaOptions.class).getPCollectionViewsStore();
+
     switch (operator.getType()) {
       case LEFT:
         final PCollectionView<Map<KeyT, Iterable<RightT>>> broadcastRight =
-            right.apply(View.asMultimap());
-        return left.apply(
+            pViews.computeViewAsMultimapIfAbsent(right, rightKeyed);
```
Comment (dmvk) [inline sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/BroadcastHashJoinTranslator.java]: I don't think comparing by pcollection is enough as we can have a different key extractor.

File: sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/EuphoriaOptions.java (hunk 2 above)
Comment (VaclavPlajt) [reply sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/EuphoriaOptions.java]: Yes, I agree. But Gradle complains during build that there is a missing setter. Is there a way of circumventing it?
The `PViewsStore` will be skipped completely. As a conclusion of mutual discussion.

File: sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/BroadcastHashJoinTranslator.java (hunk 3 above)
Comment (VaclavPlajt) [reply sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/BroadcastHashJoinTranslator.java]: Agree. But that brigs the problem of comparing lambdas for equality (again). Which we do not have good enough solution for.  So I will use `==` equality and try to highlight it in the docs.

File: sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/BroadcastHashJoinTranslator.java [hunk 4]
```diff
@@ -35,22 +38,41 @@
...
+ * </ul>
  */
 public class BroadcastHashJoinTranslator<LeftT, RightT, KeyT, OutputT>
     extends AbstractJoinTranslator<LeftT, RightT, KeyT, OutputT> {
 
+  /**
+   * Used to prevent multiple views to the same input PCollection. And therefore multiple broadcasts
+   * of the same data.
+   */
+  private Table<PCollection<?>, UnaryFunction<?, KeyT>, PCollectionView<?>> pViews =
```
[the remark this answers is not in the document]
Comment (VaclavPlajt) [reply sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/BroadcastHashJoinTranslator.java]: added

**Comment by VaclavPlajt** [comment]
Run JavaPortabilityApi PreCommit

**Pull request merged**